plugin/trino-oracle/src/main/java/io/trino/plugin/oracle/OracleClientModule.java

```diff
@@ -78,6 +80,13 @@ public static ConnectionFactory connectionFactory(BaseJdbcConfig config, Credent
                 new OracleDriver(),
```

Comment:
I think this code should be removed and the one from `JdbcModule` should be reused.

Reply:
The retry setting in this pull request is global, so if we make it disable by default for backward compatibility for all JDBC connectors except Oracle connector, it will be also disabled in Oracle connector by default which means it will be backward incompatible for Oracle connector.

Is it better to make it configurable per connector and make it enable by default only for Oracle connector?

Reply:
Is it possible to have a dedicated default configuration for Oracle connector? Or it will be backward imcompatible for Oracle Connector as retry is disabled by default.

Reply:
It has to be backward compatible. We need to allow each connector to provide its own strategy when things should be retried.

I think we could have `JdbcClient::isRetryable(SqlException)` method that could be implemented in Oracle differently.

Reply:
Is what you mean the followings?

- Making the default of `query.retry-opening-connection` true
- Returning false for all exceptions in the default retry strategy

What I'm worrying is how we can enable retry for Oracle connector and disable for other JDBC connectors by default.

> I think we could have JdbcClient::isRetryable(SqlException) method that could be implemented in Oracle differently.

This is a good idea, by the way. I will take it!

Reply:
> Returning false for all exceptions in the default retry strategy

No. By default we want to retry for `SQLTransientException`.

> What I'm worrying is how we can enable retry for Oracle connector and disable for other JDBC connectors by default.

You need to inject inject different strategy into `JdbcClient::isRetryable`.

> This is a good idea, by the way. I will take it!

There are hidden dragons :) There will be a cycle in guice as connection factory is needed for jdbc client, and now you need to inject jdbc client into connection factory.

Reply:
> There are hidden dragons :) There will be a cycle in guice as connection factory is needed for jdbc client, and now you need to inject jdbc client into connection factory.

We can

- keep it the connector responsibility to provide connector factory (no default impl of retrying and config would be bound by the connector only)
- split the `isRetryable` to a new interface, breaking `JdbcClient` "hegemony" 

(btw `isRetryable` may be different things in different context; for example insert failures need to be retried very carefully, while connection opening failures can be retried pretty blindly)

Reply:
@kokosing 

> No. By default we want to retry for SQLTransientException.

Oh, I see. You mean we can make that default for all JDBC connectors. Okay.

Reply:
>  for example insert failures need to be retried very carefully, while connection opening failures can be retried pretty blindly

@findepi This PR is only to retry connection opening. Everything else has to be handled one by one.

Reply:
> but sounds like that is not a good idea according to @findepi.

I don't read his comment this way. He suggested that we can introduce different component with possibly `optionalBInder` so by default is bound to default strategy, but other connectors like oracle can override with vendor specific retry strategy.

He also mentioned that retrying of remote query execution is non-trivial and I agree.

Reply:
> no default impl of retrying and config would be bound by the connector only

I thought this statement says we shouldn't have any default starategy for retrying which conflicts with retrying opening connection for SQLTransientException by default.

Yes, this is only for retrying opening a connection. I will wait for a while to make a decision on the direction.

Reply:
Made `RetryingConnectionCondition` optional binding. But looks like connection pooling and retry are exclusive in the current Oracle connector while this code allows to enable both connection pooling and retry.